### PR TITLE
BHBC-1593: Project EML

### PR DESCRIPTION
**Pull request by NickPhura**
# Overview

## Links to jira tickets

https://quartech.atlassian.net/browse/BHBC-1593

## This PR contains the following changes

- Add eml-service for building project eml data.
- Add eml endpoint for downloading eml data.
- Add new column to project list for downloading EML data (might need to move when new UI's are in place?)

Other:
- Fixed/tweaked a few existing tests
- Updated some mui app imports to be 1 per line

## This PR contains the following types of changes

- [ ] New feature (change which adds functionality)
- [ ] Enhancement (improvements to existing functionality)
- [ ] Bug fix (change which fixes an issue)
- [ ] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [ ] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

Please add any relevant UI screenshots, if applicable.

### CODE REVIEW

File: api/src/services/taxonomy-service.ts [hunk 1]
```diff
@@ -34,6 +34,18 @@ export class TaxonomyService {
     });
   };
 
+  async getTaxonomyFromIds(ids: number[]) {
```
Comment (shreypdev) [inline api/src/services/taxonomy-service.ts:32]: Do we need this function as we have a very similar function `getSpeciesFromIds`?

File: api/src/services/project-service.ts [hunk 2]
```diff
@@ -25,6 +25,16 @@ import { IUpdateProject } from '../paths/project/{projectId}/update';
 import { queries } from '../queries/queries';
 import { DBService } from './service';
 
+export type ProjectObject = {
```
Comment (shreypdev) [inline api/src/services/project-service.ts]: This is nice step, can we actually move this in `models/project-view.ts` or something as a interface and sanitary class which will make sure it is sent correctly everytime.

File: api/src/services/eml-service.ts [hunk 3]
```diff
@@ -0,0 +1,380 @@
+import bbox from '@turf/bbox';
```
Comment (shreypdev) [inline api/src/services/eml-service.ts:1]: Really nice and clean work on this file. I was just thinking to make the helper functions `private` methods so we dont expose those methods to initiators, will be bit cleaner while deciding what needs to be called. Also rearrage privates on top of class.

File: api/src/queries/codes/db-constant-queries.ts [hunk 4]
```diff
@@ -1,13 +1,21 @@
 import { SQL, SQLStatement } from 'sql-template-strings';
 
-export const getDbCharacterSystemConstantSQL = (constantName: string): SQLStatement =>
+export type DBSystemConstant =
```
Comment (shreypdev) [inline api/src/queries/codes/db-constant-queries.ts:3]: Nice 😎

File: api/src/paths/project/{projectId}/export/eml.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        type: 'number',
+        minimum: 1
+      },
+      required: true
+    }
+  ],
+  responses: {
+    200: {
+      description: 'Ecological Metadata Language (EML) extract production OK',
+      content: {
```
Comment (shreypdev) [inline api/src/paths/project/{projectId}/export/eml.ts:28]: This needs to be updated with returns

File: api/src/paths/project/{projectId}/export/eml.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+};
+
+export function getProjectEml(): RequestHandler {
+  return async (req, res) => {
+    defaultLog.debug({ label: 'getProjectEml', message: 'params', files: req.body });
+
+    const connection = getAPIUserDBConnection();
+
+    try {
+      const projectId = Number(req.params.projectId);
```
Comment (shreypdev) [inline api/src/paths/project/{projectId}/export/eml.ts]: NIT: like `attachments/id/list.ts`, will you be able to move this out of try. so try starts with `await connection.open();`

File: api/src/paths/project/{projectId}/export/eml.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+      await connection.open();
+
+      const emlService = new EmlService({ projectId: projectId }, connection);
```
Comment (shreypdev) [inline api/src/paths/project/{projectId}/export/eml.ts:77]: NIT: This can be `const emlService = new EmlService({ projectId }, connection);`

**Review by shreypdev — DISMISSED** [review]
Really nice job, very clean code and easy to understand and works like 💎. Just need tests and then ready to 🚀.

File: api/src/services/taxonomy-service.ts (hunk 1 above)
Comment (NickPhura) [reply api/src/services/taxonomy-service.ts:32]: The main difference is that the EML file breaks the taxonomy down into its parts, so I can't have 1 giant string.  I need the english_name and kingdom, etc, all separately.  Might be other ways to re-arrange this code, but that was just the easiest option for now.

Comment (NickPhura) [reply api/src/services/taxonomy-service.ts:32]: I can make one call the other, and just include the string formatting in one of them so the query isn't duplicated.

File: api/src/services/project-service.ts (hunk 2 above)
Comment (NickPhura) [reply api/src/services/project-service.ts]: yup, good call

File: api/src/services/eml-service.ts (hunk 3 above)
Comment (NickPhura) [reply api/src/services/eml-service.ts:1]: Yeah I can look into that.  I bounce back and forth on private a bit, as it does make it nicer to use, but it can also make it more complicated to test, and since none of this code is used by external parties, it doesn't really matter if internal code is exposed.  If we were making an NPM package or something, then definitely you'd want to not expose certain pieces.

File: api/src/paths/project/{projectId}/export/eml.ts (hunk 6 above)
Comment (NickPhura) [reply api/src/paths/project/{projectId}/export/eml.ts]: I did have it that way before, and then moved it in for w/e reason.  Will look at it

File: api/src/paths/project/{projectId}/export/eml.ts (hunk 7 above)
Comment (NickPhura) [reply api/src/paths/project/{projectId}/export/eml.ts:77]: I know, I just havent decided how I feel about that style. It is slightly less friendly to read I find. I find sometimes when I am skimming over code, that its harder to recognize the assignment.

**Review by anissa-agahchen — COMMENTED** [review]
Functionality works fine

**Review by anissa-agahchen — APPROVED** [review]
